### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by segeljakt**

### CODE REVIEW

**Comment by segeljakt** [comment]
@frej This now compiles and successfully runs all tests on my Macbook 🎉

**Comment by frej** [comment]
I started a build now, I'll tell you how it went tomorrow.

**Comment by frej** [comment]
That one worked, now testing ~1

**Comment by frej** [comment]
90b0dc3a16a49e03b2e44972207290f9fb5a523a this is needed to fix it:

```patch
diff --git a/arc-mlir/src/lib/Types.cpp b/arc-mlir/src/lib/Types.cpp
index f3b38ed..37ad89c 100644
--- a/arc-mlir/src/lib/Types.cpp
+++ b/arc-mlir/src/lib/Types.cpp
@@ -79,13 +79,13 @@ Type AppenderType::getMergeType() const { return getImpl()->mergeType; }
 Type AppenderType::parse(DialectAsmParser &parser) {
   if (parser.parseLess())
     return nullptr;
-  llvm::SMLoc loc = parser.getCurrentLocation();
+  Location loc = parser.getEncodedSourceLoc(parser.getCurrentLocation());
   mlir::Type mergeType;
   if (parser.parseType(mergeType))
     return nullptr;
   if (parser.parseGreater())
     return nullptr;
-  return AppenderType::getChecked(mergeType, loc);
+  return AppenderType::getChecked(mergeType.getContext(), mergeType, loc);
 }
```

**Comment by frej** [comment]
I have tested from 89347794535ec1fc92bc9c9bbfb227cfc58e6ca9 to f5673c8af613b3be777514c792e2f3ce8778fc7b, they all build. 5e03821ddcfbabe6457d7436e77dc46556730ab7 fails to build.

**Comment by segeljakt** [comment]
All builds pass now for me

**Comment by segeljakt** [comment]
I created `lib/arc/` in preparation for the next dialect

**Comment by segeljakt** [comment]
@frej Do you prefer keeping `.h` and `.td` files together, or in different directories? For example, keeping `.td` in `src/tblgen/`.

**Comment by frej** [comment]
> @frej Do you prefer keeping `.h` and `.td` files together, or in different directories? For example, keeping `.td` in `src/tblgen/`.

Keep them together, that's the way upstream LLVM does it.

File: arc-mlir/src/include/arc/Ops.td [hunk 1]
```diff
@@ -84,7 +84,7 @@ class ArcFloatUnaryOp<string mnemonic, list<OpTrait> traits = []>
     : ArcUnaryOpSameOperandAndResultType<mnemonic, traits>,
       Arguments<(ins FloatLike:$operand)>;
 
-def SinOp : ArcFloatUnaryOp<"sin"> {
```
Comment (frej) [inline arc-mlir/src/include/arc/Ops.td]: Why? Upstream uses this convention, so why should we be different?

File: arc-mlir/src/include/arc/Ops.td [hunk 2]
```diff
@@ -56,7 +77,7 @@ def MakeVector
   let verifier = [{ return customVerify(); }];
 }
 
-def ArcKeep : Arc_Op<"keep", []> {
```
Comment (frej) [inline arc-mlir/src/include/arc/Ops.td]: Don't drop the Arc prefix, I think the namespace for these operations is flat, so we want our own unique prefix.

File: arc-mlir/src/include/arc/Ops.td [hunk 3]
```diff
@@ -193,7 +197,7 @@ def IndexTuple : Arc_Op<"index_tuple", [NoSideEffect]> {
   let verifier = [{ return customVerify(); }];
 }
 
-def ArcBlockResult :
+def BlockResult :
```
Comment (frej) [inline arc-mlir/src/include/arc/Ops.td]: Don't drop the unique prefix

File: arc-mlir/src/include/arc/Ops.td (hunk 1 above)
Comment (segeljakt) [reply arc-mlir/src/include/arc/Ops.td]: Ok, there were some ops before without the `Op` suffix. To be consistent I can amend so every op has an `Op` suffix. @frej WDYT?

File: arc-mlir/src/include/arc/Ops.td (hunk 2 above)
Comment (segeljakt) [reply arc-mlir/src/include/arc/Ops.td]: Is there a rule for deciding which we want to have `Arc` as a prefix for?

File: arc-mlir/src/include/arc/Ops.td (hunk 1 above)
Comment (frej) [reply arc-mlir/src/include/arc/Ops.td]: Do the amend to make sure we follow the convention.

File: arc-mlir/src/include/arc/Ops.td (hunk 2 above)
Comment (frej) [reply arc-mlir/src/include/arc/Ops.td]: Looking at the other dialects, it looks like all tablegen defs has a prefix in order to not clash when a tablegen file is included.

Comment (segeljakt) [reply arc-mlir/src/include/arc/Ops.td]: True, maybe it is best to prefix everything with Arc as well?

Comment (frej) [reply arc-mlir/src/include/arc/Ops.td]: I think so

**Comment by segeljakt** [comment]
I had to make a change in the `Simplify naming scheme` commit.

**Comment by frej** [comment]
> I had to make a change in the `Simplify naming scheme` commit.

It still looks good to me :)

**Pull request merged**